File: Application/src/main/java/campus/ride/interfaces/BookingService.java
package campus.ride.interfaces;

import campus.ride.transfer.dtos.booking.BookingRequestDto;
import campus.ride.transfer.dtos.booking.BookingResponseDto;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public interface BookingService {
    CompletableFuture<BookingResponseDto> requestRide(BookingRequestDto requestDto);
    CompletableFuture<BookingResponseDto> acceptBooking(Long driveId, Long userId);
    CompletableFuture<BookingResponseDto> declineBooking(Long driveId, Long userId);
    CompletableFuture<BookingResponseDto> cancelBooking(Long driveId, Long userId);
    CompletableFuture<List<BookingResponseDto>> getBookingsByDrive(Long driveId);
    CompletableFuture<List<BookingResponseDto>> getBookingsByUser(Long userId);
    CompletableFuture<List<BookingResponseDto>> getMyBookings();
    CompletableFuture<List<BookingResponseDto>> getPendingBookingsByDrive(Long driveId);
    CompletableFuture<BookingResponseDto> getBooking(Long driveId, Long userId);
}

File: Frontend/src/app/core/services/booking.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { BookingRequest, BookingResponse } from '../models/booking.model';

@Injectable({
  providedIn: 'root'
})
export class BookingService {
  private apiUrl = `${environment.apiUrl}/bookings`;

  constructor(private http: HttpClient) {}

  requestRide(request: BookingRequest): Observable<BookingResponse> {
    return this.http.post<BookingResponse>(`${this.apiUrl}/request`, request);
  }

  acceptBooking(driveId: number, userId: number): Observable<BookingResponse> {
    return this.http.put<BookingResponse>(`${this.apiUrl}/${driveId}/${userId}/accept`, {});
  }

  declineBooking(driveId: number, userId: number): Observable<BookingResponse> {
    return this.http.put<BookingResponse>(`${this.apiUrl}/${driveId}/${userId}/decline`, {});
  }

  cancelBooking(driveId: number, userId: number): Observable<BookingResponse> {
    return this.http.put<BookingResponse>(`${this.apiUrl}/${driveId}/${userId}/cancel`, {});
  }

  getBookingsByDrive(driveId: number): Observable<BookingResponse[]> {
    return this.http.get<BookingResponse[]>(`${this.apiUrl}/drive/${driveId}`);
  }

  getBookingsByUser(userId: number): Observable<BookingResponse[]> {
    return this.http.get<BookingResponse[]>(`${this.apiUrl}/user/${userId}`);
  }

  getPendingBookingsByDrive(driveId: number): Observable<BookingResponse[]> {
    return this.http.get<BookingResponse[]>(`${this.apiUrl}/drive/${driveId}/pending`);
  }

  getBooking(driveId: number, userId: number): Observable<BookingResponse> {
    return this.http.get<BookingResponse>(`${this.apiUrl}/${driveId}/${userId}`);
  }
}
